src/semidiscretization/semidiscretization_coupled.jl

```diff
@@ -137,24 +148,27 @@ end
     @view u_ode[semi.u_indices[index]]
 end
 
+# Rcursive call of the RHS for the semidiscretizations using Lispy tuple programming
+# to avoid type instability.
+@inline function rhs!(u_ode, du_ode, t, semi, i, semi_, semi_tuple...)
+    u_loc = get_system_u_ode(u_ode, i, semi)
+    du_loc = get_system_u_ode(du_ode, i, semi)
+    rhs!(du_loc, u_loc, semi_, t)
+    if i < nsystems(semi)
+        rhs!(u_ode, du_ode, t, semi, i + 1, semi_tuple...)
+    end
+end
+
 function rhs!(du_ode, u_ode, semi::SemidiscretizationCoupled, t)
     @unpack u_indices = semi
 
     time_start = time_ns()
 
-    @trixi_timeit timer() "copy to coupled boundaries" begin
-        for semi_ in semi.semis
-            copy_to_coupled_boundary!(semi_.boundary_conditions, u_ode, semi)
-        end
-    end
+    foreach(semi_ -> copy_to_coupled_boundary!(semi_.boundary_conditions, u_ode, semi),
+            semi.semis)
 
     # Call rhs! for each semidiscretization
-    for i in eachsystem(semi)
-        u_loc = get_system_u_ode(u_ode, i, semi)
-        du_loc = get_system_u_ode(du_ode, i, semi)
-
-        @trixi_timeit timer() "system #$i" rhs!(du_loc, u_loc, semi.semis[i], t)
-    end
+    rhs!(u_ode, du_ode, t, semi, 1, semi.semis...)
```

Comment:
It appears that you need to do this (an iteration over elements with corresponding index) several times. You could simplify it by using `foreach(enumerate(semi.semis))`, but unfortunately, this causes allocations.
Instead, you can define
```julia
# Same as `foreach(enumerate(something))`, but without allocations.
#
# Note that compile times may increase if this is used with big tuples.
@inline foreach_enumerate(func, collection) = foreach_enumerate(func, collection, 1)
@inline foreach_enumerate(func, collection::Tuple{}, index) = nothing

@inline function foreach_enumerate(func, collection, index)
    element = first(collection)
    remaining_collection = Base.tail(collection)

    func((index, element))

    # Process remaining collection
    foreach_enumerate(func, remaining_collection, index + 1)
end
```
and then replace the whole function above by a simple
```suggestion
    foreach_enumerate(semi.semis) do (i, semi_)
        u_loc = get_system_u_ode(u_ode, i, semi)
        du_loc = get_system_u_ode(du_ode, i, semi)
        rhs!(du_loc, u_loc, semi_, t)
    end
```